EIPS/eip-7643.md

```diff
@@ -0,0 +1,1983 @@
+---
+eip: 7643
+title: History accumulator for pre-PoS data
+description: Commit to a single root for all block data before the merge.
+author: lightclient (@lightclient), kdeme (@kdeme)
+discussions-to: https://ethereum-magicians.org/t/eip-7643-history-accumulator-for-pre-pos-data/19014
+status: Draft
+type: Standards Track
+category: Core
+created: 2024-02-29
+---
+
+## Abstract
+
+Defines an SSZ object for accumulating all pre-PoS data and commit to the
+historical hashes accumulator's root
+`0xec8e040fd6c557b41ca8ddd38f7e9d58a9281918dc92bdb72342a38fb085e701`.
+
+## Motivation
+
+There are two main uses we consider for the historical hashes accumulator:
+
+* for users who wish to download the pre-PoS data for the execution chain and
+  verify it without executing each block, they may simply compute each block
+  hash, accumulate the epoch records, and then compare the local accumulator
+  root with the expected value.
+* additionally, the accumulator root allows for `O(log(n))` sized proofs to any
+  pre-PoS block, whereas today to achieve something similar, one must
+  recursively verify the header chain.
+
+## Specification
+
+### Historical Hashes Accumulator
+
+The historical hashes accumulator commits to the set of pre-merge headers and their
+associated total difficulty. The format for this data is defined as:
+
+```python
+EPOCH_SIZE = 8192 # blocks
+MAX_HISTORICAL_EPOCHS = 131072  # 2**17
+
+# An individual record for a historical header.
+HeaderRecord = Container[block_hash: bytes32, total_difficulty: uint256]
```

Comment:
just an idea: throw in all other header fields in there (with default values where fields didn't exist before a particular hardfork)?

Reply:
I'm really just creating an EIP around something that is already live on the portal network, so don't exactly have the ability to change the format further.

Reply:
@g11tech what's your motivation for adding the other header fields to the record?  We just need a unique record for each header and arguably even the blockhash should be sufficient.  There was some discussion in the past on possibly removing the total difficulty since it didn't really add any value for purposes of constructing the overall accumulator.

Reply:
well i was wondering about having the utility of TD, but then i thought that if TD is useful may be other fields are useful too.

but TD is an accumulator so i guess I am fine